engine/runtime-compiler/src/main/java/org/enso/compiler/pass/analyse/ImportSymbolAnalysis.java

```diff
@@ -0,0 +1,194 @@
+
+import java.util.ArrayList;
+import java.util.List;
+import org.enso.common.ScalaConversions;
+import org.enso.compiler.context.InlineContext;
+import org.enso.compiler.context.ModuleContext;
+import org.enso.compiler.core.ir.Expression;
+import org.enso.compiler.core.ir.MetadataStorage;
+import org.enso.compiler.core.ir.Module;
+import org.enso.compiler.core.ir.Name;
+import org.enso.compiler.core.ir.expression.errors.ImportExport;
+import org.enso.compiler.core.ir.module.scope.Import;
+import org.enso.compiler.data.BindingsMap;
+import org.enso.compiler.pass.IRProcessingPass;
+import org.enso.compiler.pass.MiniIRPass;
+import org.enso.compiler.pass.MiniPassFactory;
+import org.enso.compiler.pass.desugar.GenerateMethodBodies$;
+import scala.collection.immutable.Seq;
+import scala.jdk.javaapi.CollectionConverters;
+
+/**
+ * Performs analysis of `from ... import sym1, sym2, ...` statements - checks that all the symbols
+ * imported from the module can be resolved, i.e., exists. In case of unresolved symbols, replaces
+ * the IR import with {@link org.enso.compiler.core.ir.expression.errors.ImportExport}. Reports only
+ * the first unresolved symbol.
+ */
+public class ImportSymbolAnalysis implements MiniPassFactory {
+  public static final ImportSymbolAnalysis INSTANCE = new ImportSymbolAnalysis();
+
+  private ImportSymbolAnalysis() {}
+
+  @Override
+  public Seq<? extends IRProcessingPass> precursorPasses() {
+    List<IRProcessingPass> passes =
+        List.of(BindingAnalysis$.MODULE$, GenerateMethodBodies$.MODULE$);
+    return ScalaConversions.seq(passes);
+  }
+
+  @Override
+  public Seq<? extends IRProcessingPass> invalidatedPasses() {
+    return ScalaConversions.seq(List.of());
+  }
+
+  @Override
+  public MiniIRPass createForModuleCompilation(ModuleContext moduleContext) {
+    var bindingsMap = moduleContext.bindingsAnalysis();
+    return new Mini(bindingsMap);
+  }
+
+  @Override
+  public MiniIRPass createForInlineCompilation(InlineContext inlineContext) {
+    // TODO: Must not return null?
+    return null;
+  }
+
+  private static final class Mini extends MiniIRPass {
+    private final BindingsMap bindingsMap;
+
+    private Mini(BindingsMap bindingsMap) {
+      this.bindingsMap = bindingsMap;
+    }
+
+    @Override
+    public Module transformModule(Module moduleIr) {
+      for (var imp : CollectionConverters.asJava(moduleIr.imports())) {
+        var encounteredErrors = analyseSymbolsFromImport((Import.Module) imp);
+        if (encounteredErrors != null) {
+          return moduleIr.copy(
+              encounteredErrors,
+              moduleIr.exports(),
+              moduleIr.bindings(),
+              moduleIr.isPrivate(),
+              moduleIr.location(),
+              moduleIr.passData(),
+              moduleIr.diagnostics(),
+              moduleIr.id());
+        }
+      }
+      return moduleIr;
+    }
+
+    @Override
+    public Expression transformExpression(Expression expr) {
+      return expr;
```

Comment:
This is an empty, useless call that might slow the traversal down (on the instruction level). Didn't we wanted to have a way to let the `MiniIRPass` to instruct the infrastructure to skip processing of subtrees? Shouldn't this mini pass:
```java
  public MiniIRPass prepare(IR parent, Expression child) {
    return null;
  }
```
e.g. remove itself from any `Expression` processing? In such case the `transformExpression` should `throw new IllegalStateException()` as it should be never called.

Reply:
The problem is that this pass only wishes to transform `Module` IR. More specifically, it only does something useful in `transformModule`. And `transformModule` is called as the last one during the traversal - because it transforms the root node. If you would tell the traversing system that you wish to skip the rest of the iteration in any expression, you would never reach the Module root node.

Reply:
> would never reach the Module root node.

That's not how the system is supposed to work and I would be surprised if it worked that way:
- `MiniPassTraverser.compileDeep` receives `IR.Module`
- it creates `rootTask` for the `IR.Module` object
- it puts the task to `stackOfPendingIrs`
- then something happens
- at the end `rootTask.convertExpression` is called
- inside of `converExpresison` as call to `transformModule` is made